spark data preparation process, issue #3295

kangpinghuang opened the pull request:
To resolve issue  #3295, Now this pr accomplish the following tasks:
1. data etl
2. data repartition and bucket
3. rollup data build
4. write data to parquet
5. write job statistics

what are todos:
1. to support all udf that hadoop load support
2. optimize performance
3. more test cases

ref: #3433

Commit:
spark data preparation process, issue #3295

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/BitmapUnion.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public void update(MutableAggregationBuffer buffer, Row input) {
+        // here maybe there is performance problems
+        if (!input.isNullAt(0)) {
```
This performance is ensure bad, not maybe.

You could refer to `DistinctCountBufferAggregator` in Apache Druid.

kangpinghuang replied:
The performance optimization will be done in next step.

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppColumns.java:
```diff
@@ -0,0 +1,119 @@
+    public List<Class> schema;
+
+    public DppColumns() {
+        columns = null;
```
Would better init columns and schema directly to new ArrayList, not null.

kangpinghuang replied:
ok

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppColumns.java:
```diff
@@ -0,0 +1,119 @@
+import java.util.Comparator;
+
+class DppColumns implements Comparable<DppColumns>, Serializable {
+    public List<Object> columns;
```
```suggestion
    public List<Object> columns =Lists.newArrayList();
```

kangpinghuang replied:
ok

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppColumns.java:
```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public int compareTo(DppColumns other) {
+        assert (columns.size() == other.columns.size());
```
Prefer `Preconditions.checkState`

kangpinghuang replied:
ok

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppUtils.java:
```diff
@@ -0,0 +1,230 @@
+                continue;
+            }
+            if (str == null || !str.contains("=")) {
+                System.err.println("Fail to parse columnsFromPath, expected: " + columnsFromPath + ", filePath: " + filePath);
```
Please use Log. not `System.err.println`

kangpinghuang replied:
done

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/MinimalCoverRollupTreeBuilder.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Collections;
+import java.util.List;
+
+// Build RollupTree by using minimal cover strategy
```
What's the `minimal cover strategy` meaning?

kangpinghuang replied:
I add some comment and example

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:
```diff
@@ -0,0 +1,848 @@
+                        } else if (column.columnType.equals("DOUBLE")) {
+                            builder.required(PrimitiveType.PrimitiveTypeName.DOUBLE).named(column.columnName);
+                        } else {
+                            // TODO: exception handle
```
This todo should fix.

kangpinghuang replied:
done

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:
```diff
@@ -0,0 +1,848 @@
+                                                           Dataset<Row> srcDataframe,
+                                                           StructType dstTableSchema,
+                                                           EtlJobConfig.EtlFileGroup fileGroup) throws UserException {
+        System.out.println("convertSrcDataframeToDstDataframe for index:" + baseIndex);
```
Use Log

kangpinghuang replied:
done

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:
```diff
@@ -0,0 +1,848 @@
+            dppResult.fileNumber = fileNumberAcc.value();
+            dppResult.fileSize = fileSizeAcc.value();
+            dppResult.abnormalRows = abnormalRowAcc.value();
+            throw exception;
```
Are you sure you want to throw exception?

kangpinghuang replied:
yes, use exception to indicate the spark app failed.

kangkaisen commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:
```diff
@@ -0,0 +1,848 @@
+// which will make use of the distributed compute framework of spark to
+// do ETL job/sort/preaggregate/convert data format job in spark job
+// to boost the process of large amount of data load.
+public final class SparkDpp implements java.io.Serializable {
```
Please add a detail comment to describe the work flow,
Such as:
step1 : xx
step2: xxx
step3:xxx

kangpinghuang replied:
done

wangbo commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppUtils.java:
```diff
@@ -0,0 +1,230 @@
+            case "VARCHAR":
+            case "BITMAP":
+            case "OBJECT":
+                dataType = DataTypes.StringType;
```
1. For HLL and BITMAP,output schema is Binary instead of String
2. A better understand method name is,```columnTypeToClass ForInputSchema``` and ```columnTypeToDataType for output schema```

kangpinghuang replied:
the input type for hll and bitmap is string.

wangbo commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DorisRangePartitioner.java:
```diff
@@ -0,0 +1,87 @@
+        DppColumns key = (DppColumns)var1;
+        // get the partition columns from key as partition key
+        DppColumns partitionKey = new DppColumns(key, partitionKeyIndexes);
+        // TODO: optimize this by use binary search
```
maybe use a map instead of a list to avoid search is better?

wangbo commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppColumns.java:
```diff
@@ -0,0 +1,119 @@
+            } else if (columns.get(i) instanceof Date) {
+                cmp = ((Date)(columns.get(i))).compareTo((Date) (other.columns.get(i)));
+            } else {
+                cmp = ((String)(columns.get(i))).compareTo((String) (other.columns.get(i)));
```
dealing binary for bitmap and hll here

kangpinghuang replied:
will treat binary as string

yangzhg commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppUtils.java:
```diff
@@ -0,0 +1,230 @@
+        }
+        for (EtlJobConfig.EtlColumn column : columns) {
+            String columnName = column.columnName;
+            String columnType = column.columnType;
```
seems not necessary to define columnName and columnType

kangpinghuang replied:
ok, I will remove them

yangzhg commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/BitmapUnion.java:
```diff
@@ -0,0 +1,129 @@
+        try {
+            roaringBitmap.serialize(new DataOutputStream(bos));
+        } catch (IOException ioException) {
+            ioException.printStackTrace();
```
better use log

yangzhg commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/BitmapUnion.java:
```diff
@@ -0,0 +1,129 @@
+                bitmap.serialize(new DataOutputStream(bos));
+                buffer.update(0, bos.toByteArray());
+            } catch (IOException e) {
+                e.printStackTrace();
```
prefer using log instead of print to std

kangpinghuang replied:
done

kangpinghuang committed:
optimize spark dpp

1. optimize column validation logic
2. use log to replace System.err
3. add invalid rows collector
4. add some comments
5. add hll and replace udaf

kangpinghuang committed:
fix unit test

wangbo commented on fe/src/main/java/org/apache/doris/load/loadv2/dpp/DppUtils.java:
```diff
@@ -0,0 +1,252 @@
+            buffer.putLong((Long)o);
+        } else if (type.equals(DataTypes.StringType)) {
+            try {
+                String str = (String)o;
```
```suggestion
                String str = String.valueOf(o);
```
For the case，a column whose type is largeint in doris, but in source hive table,its type is bigint, type cast failed will happened here.
Or you can change the logic in ```convertSrcDataframeToDstDataframe```.
Cast source hive table column type to string when doris column type is largeint

kangpinghuang replied:
now largeint is not supported in spark dpp.

kangpinghuang committed:
fix performance problems of encoder and add atomic commit of file writer
